Factor out lti_outcome_params fixture

seanh opened the pull request:
In preparation for replacing the lti_outcome_params fixture (in a future
PR) remove some unnecessary uses of it.

This reduced coverage so I increased it again by omitting some files
that really don't need to be tested.

jon-betts commented on tests/unit/lms/views/basic_lti_launch_test.py:
```diff
@@ -52,9 +52,8 @@ def test_it_doesnt_add_report_submission_config_if_required_param_missing(
-        pyramid_request.params.update(lti_outcome_params)
         pyramid_request.params.update(
             {"tool_consumer_info_product_family_code": "whiteboard"}
         )
```
This is a little OTT for setting a single value :woman_shrugging:

seanh replied:
Fixed

jon-betts commented on .coveragerc:
```diff
@@ -6,9 +6,11 @@ source =
 omit =
     lms/scripts/*
+    lms/extensions/feature_flags/views/test.py
+    lms/views/feature_flags_test.py
```
Why are these files not subject to coverage? A comment here would be nice, otherwise it just lives in heads and is hard to question.

seanh replied:
They're just weird little test views for the feature flags mechanism. They're not anything actually being used by users. They could even just be deleted but they can be handy now and then

seanh replied:
Added a comment

jon-betts commented on .coveragerc:
```diff
@@ -6,9 +6,11 @@ source =
 show_missing = True
 precision = 2
-fail_under = 98.49
+fail_under = 98.98
```
If we are dropping covering in this test, is it worth just fessing up?

seanh replied:
Naah, let us always move towards 100%. Once we hit 100% it will no longer be possible for a PR to reduce coverage by deleting more lines than it adds and stuff like that, so things will be a little more convenient

jon-betts commented on tests/unit/lms/views/basic_lti_launch_test.py:
```diff
@@ -248,16 +247,8 @@ def test_it_configures_via_url(
-        lti_outcome_params,
-        via_url,
-        ModuleItemConfiguration,
+        self, context, pyramid_request, frontend_app, via_url, ModuleItemConfiguration,
```
I freaking love it when this happens!

seanh commented:
I've added some new commits to this PR:

* Increase test coverage by adding tests for `views/application_instances.py`

* Remove the `lti_outcome_params` fixture. The `"lis_result_sourcedid"` and `"lis_outcome_service_url"` params that it added have been added to a general `pyramid_request` fixture in the module since all launch requests actually have these standard params.

  Certain specific requests -- e.g. a launch by a teacher, or a SpeedGrader launch -- are missing one or another of these params. Tests for those specific cases delete those params from the request.

  These weren't added to the `pyramid_request` in `conftest.py` because doing so breaks some unrelated tests for reasons I don't want to deal with now.

* The `"tool_consumer_info_product_family_code"` that `lti_outcome_params` added has been added to the universal `pyramid_request` in `conftest.py`. Every launch request has this. It's set to `"whiteboard"` (i.e. not Canvas) by default.

* Added a `canvas_request` fixture that sets `"tool_consumer_info_product_family_code"` to `"canvas"`, and changed tests for Canvas-specific behaviour to use it

seanh commented on lms/views/application_instances.py:
```diff
@@ -10,18 +10,11 @@
-
-    if not developer_key or not developer_secret:
-        developer_key = None
-        developer_secret = None
```
Amusing code here

seanh commented on lms/views/application_instances.py:
```diff
@@ -10,15 +10,14 @@
-        developer_key = None
-
-    if developer_secret == "":
-        developer_secret = None
```
If either of these is falsey (e.g. an empty string) then they both get set to `None` right below, so these four lines are unnecessary. (The fact that it silently deletes the given value for either one of these fields if the other field is empty doesn't seem great but I'm not fixing this page right now, just adding test coverage.)

jon-betts commented on tests/unit/lms/views/test_application_instance.py:
```diff
@@ -1,25 +1,68 @@
+
+        ai = pyramid_request.db.query(ApplicationInstance).one()
+        assert ai.developer_key is None
+        assert ai.developer_secret is None
```
These two tests look like they could be a parameterised `test_if_developer_key_or_secret_is_missing_the_other_isnt_saved` or something by iterating over the key to change.

seanh replied:
Done

jon-betts commented on tests/unit/lms/views/test_application_instance.py:
```diff
@@ -1,25 +1,68 @@
-            .filter(ApplicationInstance.lms_url == pyramid_request.params["lms_url"])
-            .count()
-        )
-        assert new_count == initial_count + 1
```
I'm assuming this isn't necessary because we truncate the DB now?

seanh replied:
The unit tests always truncated the DB. Whoever wrote this just didn't know what they were doing
